Support using the result as a promise

jkrems opened the pull request:
Adds additional properties to the result of `hub.fetch` that make it usable as a promise. Docs will follow if we agree to do this.

Allows for things like:

```js
github.emoji().then(function(emoji) { console.log(emoji); });

Promise.all([
  github.emoji()
  github.repo('groupon/gofer').contents()
]).spread(function(emoji, rootContents) {
  // fun with emoji and the root contents of gofer
});

async function evenMoreFun() {
  try {
    const [ emoji, rootContents ] = await Promise.all([
      github.emoji(), github.repo('groupon/gofer').contents()
    ]);
    if (rootContents.some(entry => entry.name === 'README.md')) {
      console.log(emoji);
    }
  } catch (fetchError) {
    // oooh yeah
  }
}
```

abloom commented on src/gofer.coffee:
```diff
@@ -170,12 +175,12 @@ class Gofer
 
-    @hub.fetch options, (err, body, response, responseData) ->
-      parseJSON = options.parseJSON ? isJsonResponse(response, body)
-      return cb err, body, responseData, response unless parseJSON
```
are you losing the ability to parse JSON with a callback?

jkrems replied:
Nope, that moved to `hub.coffee` below. Made it easier to reason about.

abloom commented on src/gofer.coffee:
```diff
@@ -39,9 +39,14 @@ Hub = require './hub'
-{ safeParseJSON, isJsonResponse } = require './json'
 {extend} = require 'lodash'
 
+promiseHelpers =
```
this looks unused

abloom commented on src/hub.coffee:
```diff
@@ -38,13 +38,37 @@ HRDuration = require 'hrduration'
+    value: (onError, onSuccess) ->
+      @asPromise().then onError, onSuccess
+
+  getBody:
```
why build `getBody` and `getResponse`?

abloom commented on src/hub.coffee:
```diff
@@ -119,12 +146,21 @@ module.exports = Hub = ->
+    if typeof done == 'function'
+      req.on RESULT_EVENT, done
+
+    return Object.defineProperties req, promiseHelpers(Hub.Promise)
```
Why aren't you calling `asPromise` here? Why create those helpers? Why not just let it resolve to:

````coffeescript
myApi.query().then (data) -> 
  data === [body, response, responseData]
````

jkrems replied:
Because 99% of cases at least I don't want that array. There's so much noise all over code using `gofer` + `async` just to get rid of that array. I'd be fine with `.then (body) ->`. That'd make more sense imo. But then there's got to be another way of getting the response itself.

jkrems commented:
I added a longer comment why `.then` exists and why there's a second, more official API (`getBody`) to achieve the same thing. Also removed the unused `promiseHelpers` in `gofer.coffee`.

abloom commented on test/promise_helpers.test.coffee:
```diff
@@ -0,0 +1,90 @@
+      assert.equal 'not found', err.body.message
+
+  it 'exposes the parsed response body', ->
+    myApi.query().getBody().then (reqMirror) ->
```
The think that kept me from giving this a :shipit: last night was wondering how it handled errors. Can you confirm this would be legitimate:

````
myApi.fail().getBody().then((bodyData) -> 
  console.log(bodyData)
).catch((err) -> 
  console.log(err)
)
````

You're right, most of the time I don't care about the status, I just want the body, or I want to be alerted to a failure.

jkrems replied:
Yep, errors can be properly caught. In io.js:

```
co ->
  try
    console.log yield myApi.fail().getBody()
  catch err
    console.log error
```

jkrems commented:
Added docs and moved the promise helpers into their own file. I think this is ready for final review.

abloom commented on API.md:
```diff
@@ -97,7 +97,23 @@ Useful if you know that you'll need custom timeouts for this one call or you wan
+* `getResponse()`: A promise of the response object.
+* `then(onResolved, onRejected)`: This function exists for convenience only.
+  It makes it possible to use the result as a thenable
+  than can be passed into `Promise.resolve`, `Promise.all`, etc..
```
"**that** can be passed into"

abloom commented:
one typo in the docs. fix it and :shipit:

jkrems commented:
v2.3.0
